src/org/skypro/skyshop/basket/ProductBasket.java

```diff
@@ -2,67 +2,91 @@
 
-    public Product[] getBasket() {
+    public HashMap<String, ArrayList<Product>> getBasket() {
         return basket;
     }
 
     public void addProduct(Product newProduct) {
-        for (int i = 0; i < maxProduct; i++) {
-            if (basket[i] == null) {
-                basket[i] = newProduct;
-                return;
-            }
-        }
-        System.out.println("Невозможно добавить продукт");
+        ArrayList<Product> productList = new ArrayList<>();
+        productList.add(newProduct);
+        basket.put(newProduct.getName(), productList);
     }
 
     public int sumPrice() {
         int sum = 0;
-        for (int i = 0; i < quantityProduct(); i++) {
-            sum += basket[i].getPrice();
-        }
-        return sum;
-    }
-
-    public int quantityProduct() {
-        int quantity = 0;
-        for (int i = 0; i < maxProduct; i++) {
-            if (basket[i] != null) {
-                quantity++;
+        for (String kay : basket.keySet()) {
+            for (int i = 0; i < basket.get(kay).size(); i++) {
+                sum += basket.get(kay).get(i).getPrice();
             }
         }
-        return quantity;
+        return sum;
     }
 
     public void printContentsBasket() {
-        if (quantityProduct() == 0) {
+        if (basket.isEmpty()) {
             System.out.println("в корзине пусто");
             return;
         }
-        for (int i = 0; i < quantityProduct(); i++) {
-            System.out.println(basket[i].toString());
+        for (String kay : basket.keySet()) {
+            for (int i = 0; i < basket.get(kay).size(); i++) {
+                System.out.println(basket.get(kay).get(i).toString());
+            }
         }
         System.out.println("Итого: " + sumPrice());
+        System.out.println("Специальных товаров: " + calculateSpecial());
     }
 
     public boolean checkProduct(String product) {
-        for (int i = 0; i < quantityProduct(); i++) {
-            if (product.equals(basket[i].getName())) {
-                return true;
+        if (basket.isEmpty()) {
+            return false;
+        }
+        for (String kay : basket.keySet()) {
+            for (int i = 0; i < basket.get(kay).size(); i++) {
+                if (product.equals(basket.get(kay).get(i).getName())) {
+                    return true;
+                }
             }
         }
         return false;
     }
 
     public void clearBasket() {
-        for (int i = 0; i < maxProduct; i++) {
-            basket[i] = null;
+        basket.clear();
+    }
+
+    public int calculateSpecial() {
+        int sum = 0;
+        for (String kay : basket.keySet()) {
+            for (int i = 0; i < basket.get(kay).size(); i++) {
+                if (basket.get(kay).get(i).isSpecial()) {
+                    sum++;
+                }
+            }
         }
+        return sum;
     }
 
+    public List removeProduct(String delete) {
```

Comment:
1) Возвращать нужно параметризированный List

```suggestion
    public List<Product> removeProduct(String delete) {
```

2) У мапы есть уже метод для удаления значения по ключу, нужно использовать его